# Array-Oriented Programming with NumPy


<table align="left">
  <td>
    <a href="https://colab.research.google.com/github/phonchi/nsysu-math105A/blob/master/static_files/presentations/03_Basic_structure.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>
  </td>
  <td>
    <a target="_blank" href="https://kaggle.com/kernels/welcome?src=https://github.com/phonchi/nsysu-math105A/blob/master/static_files/presentations/03_Basic_structure.ipynb"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" /></a>
  </td>
</table>

The  `NumPy`  (Numerical  Python)  library first appeared in  2006  and is the preferred Python array implementation. It offers a high-performance, richly functional n-dimensional array type called `array,` which, from this point forward, we’ll refer to by its synonym, `array.` Operations on arrays are up to one or two orders of magnitude faster than those on lists. Many Python libraries depend on `NumPy.` Many popular data science libraries such as Pandas, SciPy (Scientific Python), and Keras (for deep learning) are built on or depend on NumPy.

In this chapter, we explore the array’s basic capabilities. The built-in lists can have multiple dimensions. You generally process multi-dimensional lists with nested loops or list comprehensions with multiple clauses. A strength of `NumPy` is “array-oriented programming,” **which uses functional-style programming with internal iteration to make array manipulations concise and straightforward**, eliminating the kinds of bugs that can occur with the external iteration of explicitly programmed loops. 

In `Python` the types are dynamically inferred and we do not have to allocate the memory by ourselves. This type of flexibility also points to the fact that `Python` variables are more than just their values; they also contain extra information about the type and the size of the value:

<center><img src="./cint_vs_pyint.png"></center>
<div align="center"> source: https://jakevdp.github.io/PythonDataScienceHandbook/figures/cint_vs_pyint.png </div>

Similarly, the `list` in `Python` is very flexible that can store heterogeneous objects. But this flexibility comes at a cost: to allow these flexible types, each item in the list must contain its type, size, and other information. Every element is a complete `Python` object. In the special case that all variables are of the same type, **much of this information is redundant**, so storing the data in a fixed-type array can be much more efficient. The difference between a dynamic-type list and a fixed-type (`NumPy`-style) array is illustrated:

<center><img src="array_vs_list.png"></center>
<div align="center"> source: https://jakevdp.github.io/PythonDataScienceHandbook/figures/array_vs_list.png </div>

At the implementation level, the `array` essentially contains a single pointer to one contiguous block of data. The `Python` `list`, on the other hand, includes a pointer to a block of pointers, each of which in turn points to a whole `Python` object like the `Python` integer we saw earlier. 

> The advantage of the `list` is flexibility: because each list element is a full structure containing both data and type information, the list can be filled with data of any desired type. Fixed-type `NumPy`-style arrays lack this flexibility but are much more efficient for storing and manipulating data.

From the previous lecture, we know that every object consists of data and methods. The `ndarray` object of the `NumPy` package not only provides efficient storage of array-based data but adds to this **efficient operations** on that data. 

## Creating  `array` from Existing Data 

The `NumPy` documentation recommends importing the `numpy` module as `np` so that you can access its members with "`np.`"

In [1]:
import numpy as np

The `numpy` module provides various functions for creating arrays. Here we use the `array()` function, which receives a collection of elements and returns a new array containing the argument's elements. Let’s pass a `list` for example: 

In [4]:
numbers = np.array([2, 3, 5, 7, 11])
numbers, type(numbers)

(array([ 2,  3,  5,  7, 11]), numpy.ndarray)

The `array()` function copies its argument's contents into the `array`. Note that the type is `numpy.ndarray`, but all arrays are output as "array."

### Multidimensional Arguments

The `array()` function copies its argument's dimensions. Let's create an `array` from a two-row-by-three-column `list`:

In [5]:
np.array([[1, 2, 3], [4, 5, 6]])

array([[1, 2, 3],
       [4, 5, 6]])

`NumPy` auto-formats arrays based on their number of dimensions, aligning the columns within each row.

#### `array`  Attributes 

The `array` function determines an array's element type from its argument's elements. You can check the element type with an array's `dtype` attribute:

In [ ]:
integers = np.array([[1, 2, 3], [4, 5, 6]])
floats = np.array([0.0, 0.1, 0.2, 0.3, 0.4])

integers.dtype, floats.dtype

(dtype('int32'), dtype('float64'))

As you’ll see in the next section, various array-creation functions receive a `dtype` keyword argument so you can specify an array’s element type. 

For performance reasons, `NumPy` is written in the C programming language and uses C's data types. By default, `NumPy` stores integers as the `NumPy` type `int_` values — which correspond to 32-bit (4-byte) integers in C (this may be platform-dependent) — and stores floating-point numbers as the NumPy type `float64` values — which correspond to 64-bit (8-byte) floating-point values (double) in C. In our examples, most commonly, you'll see the types `int64`, `float64` and `bool` for non-numeric data (such as strings). The complete list of supported types is at [https://docs.scipy.org/doc/numpy/user/basics.types.html](https://docs.scipy.org/doc/numpy/user/basics.types.html). 

The attribute `ndim` contains an array's number of dimensions and the attribute `shape` contains a tuple specifying an array's dimensions: 

In [9]:
print(integers.ndim)
print(floats.ndim)

2
1


In [10]:
print(integers.shape)
print(floats.shape)

(2, 3)
(5,)


Here, integers have 2 rows and 3 columns (6 elements) and floats are one-dimensional, containing 5 floating numbers.

You can view an array’s total number of elements with the attribute `size` and the number of bytes required to store each element with `itemsize`:

In [12]:
print(integers.size)
print(integers.itemsize)
print(floats.size)
print(floats.itemsize)

6
4
5
8


Note that the integers' size is the product of the shape tuple's values — two rows of three elements each for a total of six elements. In each case, `itemsize` is 4 because integers contain `int32` values and 8 since floats contain `float64` values.

### Filling `array` with Specific Values 

`NumPy` provides functions `zeros()`, `ones()` and `full()` for creating arrays containing 0s, 1s or a specified value, respectively. By default, `zeros()` and `ones()` create arrays containing `float64` values. We’ll show how to customize the element type momentarily. The first argument to these functions must be an integer or a tuple of integers specifying the desired dimensions. For an integer, each function returns a one-dimensional array with the specified number of elements:

In [ ]:
np.zeros(5)

array([0., 0., 0., 0., 0.])

For a tuple of integers, these functions return a multidimensional array with the specified dimensions. You can specify the array's element type with the `zeros()` and `ones()` function’s `dtype` keyword argument:

In [ ]:
np.ones((2, 4), dtype=np.int64)

array([[1, 1, 1, 1],
       [1, 1, 1, 1]], dtype=int64)

The array returned by `full()` contains elements with the second argument's value and type: 

In [15]:
np.full((3, 5), 13)

array([[13, 13, 13, 13, 13],
       [13, 13, 13, 13, 13],
       [13, 13, 13, 13, 13]])

### Creating `array` from `Ranges` 

#### Creating Integer Ranges with `arange()` 

Let's use `NumPy`'s `arange` function to create integer ranges — similar to using the built-in function `range`. In each case, `arange` first determines the resulting array’s number of elements, allocates the memory, then stores the specified range of values in the array: 

In [16]:
np.arange(5)

array([0, 1, 2, 3, 4])

In [17]:
np.arange(5, 10)

array([5, 6, 7, 8, 9])

In [18]:
np.arange(10, 1, -2) 

array([10,  8,  6,  4,  2])

> It is the same as `range()` which takes three arguments `numpy.arange(start, stop, step)`

#### Creating Floating-Point Ranges with `linspace()`

You can produce evenly spaced floating-point ranges with `NumPy`'s `linspace()` function. The function’s first two arguments specify the starting and ending values in the range, **and the ending value is included in the array**. The optional keyword argument `num` specifies the number of evenly spaced values to produce:

In [19]:
np.linspace(0.0, 1.0, num=5)

array([0.  , 0.25, 0.5 , 0.75, 1.  ])

#### Reshaping an `array` 

You also can create an `array` from a range of elements, then use the array method `reshape()` to transform the one-dimensional array into a multidimensional array. Let's create an `array` containing the values from 1 through 20, then reshape it into four rows by five columns:

In [20]:
np.arange(1, 21).reshape(4, 5)

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20]])

Note the ***chained method*** calls in the preceding snippet. First, `arange` produces an array containing the values 1–20. Then we call `reshape()` on that array to get the 4-by-5 array that was displayed. You can `reshape()` any array, provided that the new shape has the same number of elements as the original. So a six-element one-dimensional array can become a 3-by-2 or 2-by-3 array, and vice versa! 

When displaying an `array`, if there are many items, `NumPy` drops the middle rows, columns or both from the output. The following snippets generate 100,000 elements.

In [ ]:
np.arange(1, 100001).reshape(100, 1000)

array([[     1,      2,      3, ...,    998,    999,   1000],
       [  1001,   1002,   1003, ...,   1998,   1999,   2000],
       [  2001,   2002,   2003, ...,   2998,   2999,   3000],
       ...,
       [ 97001,  97002,  97003, ...,  97998,  97999,  98000],
       [ 98001,  98002,  98003, ...,  98998,  98999,  99000],
       [ 99001,  99002,  99003, ...,  99998,  99999, 100000]])

The above case shows the first and last three of the 3 rows, and the first and last three of the 3000 columns. The notation `...` represents the missing data.

### `List` vs. `array`  Performance: Introducing  `%timeit`  

Most `array` operations execute significantly faster than corresponding `list` operations. To demonstrate, we’ll use the `%timeit` magic command, which times the average duration of operations. 

In [29]:
import random

Here, let’s use the `random` module’s `randint()` function with a list comprehension to create a list of six million die rolls and time the operation using `%timeit`:

In [31]:
%timeit rolls_list = [random.randint(1, 6) for i in range(0, 6_000_000)] #_ is use to separate long integer

2.19 s ± 29.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


> By default, `%timeit` executes a statement in a loop, and it runs the loop seven times. If you do not indicate the number of loops, `%timeit` chooses an appropriate value.

Now, let’s use the `randint()` function from the `numpy.random` module to create an array

In [32]:
%timeit rolls_array = np.random.randint(1, 7, 6_000_000)

43 ms ± 872 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


## Indexing and Slicing 

One-dimensional arrays can be indexed and sliced using the same syntax and techniques demonstrated in the "Lists and Tuples" chapter. Here, we focus on array-specific indexing and slicing capabilities. 

To select an element in a two-dimensional array, specify a tuple containing the element's row and column indices in square brackets:

In [54]:
grades = np.array([[87, 96, 70], [100, 87, 90],
                   [94, 77, 90], [100, 81, 82]])
grades

array([[ 87,  96,  70],
       [100,  87,  90],
       [ 94,  77,  90],
       [100,  81,  82]])

In [55]:
grades[0, 1]  # row 0, column 1

96

To select a single row, specify only one index in square brackets:

In [56]:
grades[1]

array([100,  87,  90])

To select multiple sequential rows, use slice notation:

In [57]:
grades[0:2]

array([[ 87,  96,  70],
       [100,  87,  90]])

To select multiple non-sequential rows, use a list of row indices (fancy indexing):

In [58]:
grades[[1, 3]]

array([[100,  87,  90],
       [100,  81,  82]])

You can select subsets of the columns by providing a tuple specifying the row(s) and column(s) to select. Each can be a specific index, a slice or a list. Let’s select only the elements in the first column: 

In [59]:
grades[:, 0]

array([ 87, 100,  94, 100])

The 0 after the comma indicates that we’re selecting only column 0. The `:` before the comma indicates which rows within that column to select. **In this case, `:` is a slice representing all rows**. You can select consecutive columns using a slice:

In [60]:
grades[:, 1:3]

array([[96, 70],
       [87, 90],
       [77, 90],
       [81, 82]])

or specific columns using a list of column indices:

In [61]:
grades[:, [0, 2]]

array([[ 87,  70],
       [100,  90],
       [ 94,  90],
       [100,  82]])

### Views: Shallow Copies

***Views*** are objects "see" the data in other objects, rather than having their own copies of the data. Views are also known as ***shallow copies***. Various `array` methods and slicing operations produce views of an array's data. The `array` method `view()` returns a new array object with a view of the original array object's data. First, let’s create an array and a view of that array:

In [76]:
numbers = np.arange(1, 6)
numbers2 = numbers.view()

We can use the built-in `id()` function to see that `numbers` and `numbers2` are different objects:

In [77]:
id(numbers), id(numbers2)

(2050186483472, 2050186488944)

**To prove that `numbers2` views the same data as `numbers`**, let’s modify an element in `numbers`, then display both arrays:

In [78]:
numbers[1] *= 10
numbers

array([ 1, 20,  3,  4,  5])

In [79]:
numbers2

array([ 1, 20,  3,  4,  5])

Similarly, changing a value in the view also changes that value in the original array:

In [80]:
numbers2[1] /= 10
numbers, numbers2

(array([1, 2, 3, 4, 5]), array([1, 2, 3, 4, 5]))

Slices also create views. Let’s make `numbers2` a slice that views only the first three elements of numbers:

In [81]:
numbers2 = numbers[0:3]
numbers2

array([1, 2, 3])

Again, we can confirm that `numbers` and `numbers2` are different objects with `id()`: 

In [82]:
id(numbers), id(numbers2)

(2050186483472, 2050186489712)

Now, let’s modify an element both arrays share, then display them. Again, we see that `numbers2` is a view of `numbers`:

In [83]:
numbers[1] *= 20
numbers

array([ 1, 40,  3,  4,  5])

In [84]:
numbers2

array([ 1, 40,  3])

### Deep Copies

Though views are separate `array` objects, they save memory by sharing element data from other arrays. However, when sharing mutable values, sometimes creating a ***deep copy*** with independent copies of the original data is necessary. This is especially important in multi-core programming, where separate parts of your program could attempt to modify your data at the same time, possibly corrupting it. 

The `array` method `copy()` returns a new array object with a deep copy of the original array object’s data. First, let’s create an array and a deep copy of that array:

In [85]:
numbers = np.arange(1, 6)
numbers2 = numbers.copy()

To prove that `numbers2` has a separate copy of the data in `numbers`, let’s modify an element in `numbers`, then display both arrays: 

In [86]:
numbers[1] *= 10
numbers

array([ 1, 20,  3,  4,  5])

In [87]:
numbers2

array([1, 2, 3, 4, 5])

>  Recall that if you need deep copies of other types of `Python` objects, pass them to the `copy` module’s `deepcopy()` function. 

### Reshaping and Transposing 

We've used `array` method `reshape()` to produce two-dimensional arrays from one-dimensional ranges. `NumPy` provides various other ways to reshape arrays.

The array methods `reshape()` and `resize()` both enable you to change an array's dimensions. Method `reshape()` returns a view (shallow copy) of the original array with the new dimensions. It does not modify the original array:

In [2]:
grades = np.array([[87, 96, 70], [100, 87, 90]])
grades

array([[ 87,  96,  70],
       [100,  87,  90]])

In [3]:
grades.reshape(1, 6)

array([[ 87,  96,  70, 100,  87,  90]])

In [4]:
grades

array([[ 87,  96,  70],
       [100,  87,  90]])

A common trick is that you can use `-1` to specify the shape in `resahpe()`. The length of the dimension set to -1 is automatically determined by inferring from the specified values of other dimensions:

In [5]:
grades.reshape(-1, 3) # Same as grades.reshape(2, 3)

array([[ 87,  96,  70],
       [100,  87,  90]])

Method `resize()` modifies the original array’s shape:

In [6]:
grades.resize(1, 6)
grades

array([[ 87,  96,  70, 100,  87,  90]])

You can take a multidimensional array and flatten it into a single dimension with the methods `flatten()` and `ravel()`. 

Method `flatten()` deep copies the original array's data:

In [93]:
grades = np.array([[87, 96, 70], [100, 87, 90]])
grades

array([[ 87,  96,  70],
       [100,  87,  90]])

In [94]:
flattened = grades.flatten()
flattened

array([ 87,  96,  70, 100,  87,  90])

In [95]:
grades

array([[ 87,  96,  70],
       [100,  87,  90]])

Method `ravel()` produces a view of the original array, which shares the grades array's data!

In [96]:
raveled = grades.ravel()
raveled

array([ 87,  96,  70, 100,  87,  90])

In [97]:
raveled[0] = 100
grades

array([[100,  96,  70],
       [100,  87,  90]])

You can quickly transpose an array's rows and columns, so the rows become the columns and the columns become the rows. The `T` attribute returns a transposed view (shallow copy) of the array. The original `grades` array represents two students' grades (the rows) on three exams (the columns). Let's transpose the rows and columns to view the data as the grades on three exams (the rows) for two students (the columns):

In [98]:
grades.T

array([[100, 100],
       [ 96,  87],
       [ 70,  90]])

Transposing does not modify the original array:

In [99]:
grades

array([[100,  96,  70],
       [100,  87,  90]])

You can combine arrays by adding more columns or more rows — known as horizontal stacking and vertical stacking. Let's create another 2-by-3 array of grades:

In [101]:
grades2 = np.array([[94, 77, 90], [100, 81, 82]])
grades2

array([[ 94,  77,  90],
       [100,  81,  82]])

Let's assume `grades2` represents three additional exam grades for the two students in the `grades` array. We can combine `grades` and `grades2` with `NumPy`'s `hstack()` (horizontal stack) function by passing a `tuple` containing the arrays to combine. The extra parentheses are required because `hstack()` expects one argument:

In [102]:
np.hstack((grades, grades2))

array([[100,  96,  70,  94,  77,  90],
       [100,  87,  90, 100,  81,  82]])

Next, let's assume that `grades2` represents two more students' grades on three exams. In this case, we can combine `grades` and `grades2` with `NumPy`'s `vstack()` (vertical stack) function: 

In [103]:
np.vstack((grades, grades2))

array([[100,  96,  70],
       [100,  87,  90],
       [ 94,  77,  90],
       [100,  81,  82]])

### `NumPy` Calculation Methods

An `array` has various methods that perform calculations using its contents. **By default, these methods ignore the array's shape and use all the elements in the calculations.** For example, calculating the mean of an array totals all of its elements regardless of its shape, then divides by the total number of elements. **You can perform these calculations on each dimension as well.** For example, in a two-dimensional array, you can calculate each row's mean and each column's mean. 

In [45]:
grades = np.array([[87, 96, 70], [100, 87, 90],
                    [94, 77, 90], [100, 81, 82]])
grades

array([[ 87,  96,  70],
       [100,  87,  90],
       [ 94,  77,  90],
       [100,  81,  82]])

We can use methods to calculate `sum()`, `min()`, `max()`, `mean()`, `std()` (standard deviation) and `var()` (variance) — each is a functional-style programming reduction:

In [46]:
print(grades.sum())
print(grades.min())
print(grades.max())
print(grades.mean())
print(grades.std())
print(grades.var())

1054
70
100
87.83333333333333
8.792357792739987
77.30555555555556


#### Calculations by Row or Column

Many calculation methods can be performed on specific array dimensions, known as the array’s ***axes***. These methods receive an `axis` keyword argument that specifies which dimension to use in the calculation, giving you a quick way to perform calculations by row or column in a two-dimensional array. 

Assume that you want to calculate the average grade on each exam, represented by the columns of `grades`. Specifying `axis=0` performs the calculation on all the row values within each column:

In [47]:
grades.mean(axis=0)

array([95.25, 85.25, 83.  ])

So 95.25 above is the average of the first column’s grades (87, 100, 94 and 100), 85.25 is the average of the second column’s grades (96, 87, 77 and 81) and 83 is the average of the third column’s grades (70, 90, 90 and 82). Similarly, specifying `axis=1` performs the calculation on all the column values within each individual row. To calculate each student’s average grade for all exams, we can use:

<center><img src="https://scipy-lectures.org/_images/reductions.png"></center>
<div align="center"> source: https://scipy-lectures.org/_images/reductions.png </div>

In [48]:
grades.mean(axis=1)

array([84.33333333, 92.33333333, 87.        , 87.66666667])

This produces four averages—one each for the values in each row. So 84.33333333 is the average of row 0’s grades (87, 96 and 70), and the other averages are for the remaining rows. See [https://numpy.org/doc/stable/reference/arrays.ndarray.html](https://numpy.org/doc/stable/reference/arrays.ndarray.html) for more methods.

> For more operations such as linear algebra, you can use the sub-module `numpy.linalg`, which implements basic linear algebra, such as solving linear systems, singular value decomposition, etc. However, it is not guaranteed to be compiled using efficient routines, and thus we recommend the use of `scipy.linalg`, which will introduce in a later chapter.

## `array`  Operators

`NumPy` provides many operators which enable you to write simple expressions that perform operations on **entire arrays**. First, let's perform **element-wise arithmetic with arrays and numeric values** by using arithmetic operators and augmented assignments. Element-wise operations are applied to every element, so the snippet below multiplies every element by 2 and cubes every element. Each returns a new array containing the result: 

In [8]:
numbers = np.arange(1, 6) # array([1,2,3,4,5])
numbers * 2

array([ 2,  4,  6,  8, 10])

In [36]:
numbers ** 3

array([  1,   8,  27,  64, 125], dtype=int32)

Augmented assignments modify every element in the left operand in place!

In [37]:
numbers += 10
numbers

array([11, 12, 13, 14, 15])

### Broadcasting 

Normally, the arithmetic operations require as operands two arrays of the same size and shape. When one operand is a single value, called a scalar, `NumPy` performs the elementwise calculations **as if the scalar were an array of the same shape as the other operand, but with the scalar value in all its elements.** This is called ***broadcasting***. Snippets above use this capability. For example, `numbers * 2` is equivalent to `numbers * [2, 2, 2, 2, 2]`

Broadcasting also can be applied between `arrays` of different sizes and shapes, enabling some concise and powerful manipulations. We’ll show more examples of broadcasting later in the chapter when we introduce `NumPy`'s universal functions.

#### Arithmetic Operations Between arrays 

You may perform arithmetic operations and augmented assignments between arrays of the **same shape**. Let’s multiply the one-dimensional arrays `numbers `and `numbers2` (created below) that each contains five elements: 

In [40]:
numbers2 = np.linspace(1.1, 5.5, 5) # array([ 1.1,  2.2,  3.3,  4.4,  5.5])
numbers * numbers2

array([12.1, 26.4, 42.9, 61.6, 82.5])

The result is a new array formed by multiplying the arrays element-wise in each operand — `11 * 1.1, 12 * 2.2, 13 * 3.3`, etc. Arithmetic between arrays of integers and floating-point numbers results in an array of floating-point numbers. 

In [106]:
c = np.ones((3, 3))
c * c 

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

Note that the above operation is not matrix multiplication. To perform matrix multiplication use the `dot()` method!

In [107]:
c.dot(c)

array([[3., 3., 3.],
       [3., 3., 3.],
       [3., 3., 3.]])

We can similarly extend broadcasting to arrays of higher dimensions. Observe the result when we add a one-dimensional array to a two-dimensional array:

In [108]:
a = np.array([0, 1, 2])
M = np.ones((3, 3))
M + a

array([[1., 2., 3.],
       [1., 2., 3.],
       [1., 2., 3.]])

Here the one-dimensional array `a` is stretched, or broadcasted, across the second dimension in order to match the shape of `M`.

#### Rules of Broadcasting

Broadcasting in `NumPy` follows a strict set of rules to determine the interaction between the two arrays:

- Rule 1: If the two arrays differ in their number of dimensions, the shape of the one with fewer dimensions is padded with ones on its leading (left) side.
- Rule 2: If the shape of the two arrays does not match in any dimension, the array with a shape equal to 1 in that dimension is stretched to match the other shape.
- Rule 3: If in any dimension, the sizes disagree and neither is equal to 1, an error is raised.

<center><img src="https://scipy-lectures.org/_images/numpy_broadcasting.png"></center>
<div align="center"> source: https://scipy-lectures.org/_images/numpy_broadcasting.png </div>

Now let's take a look at an example where both arrays need to be broadcast:

In [109]:
a = np.arange(3).reshape((3, 1))
b = np.arange(3)
a + b

array([[0, 1, 2],
       [1, 2, 3],
       [2, 3, 4]])

We'll start by determining the shapes of the arrays:

- a.shape is (3,1)
- b.shape is (3,) 

Rule 1 says we must pad the shape of  `b`  with ones:

- a.shape is (3,1)
- b.shape becomes (1,3) 

And rule 2 tells us that we must upgrade each of these 1s to match the corresponding size of the other array:

- a.shape is (3,3)
- b.shape becomes (3,3) 

Now because the results match, these shapes are compatible!

Next, let's look at an example in which the two arrays are incompatible!

In [112]:
M = np.ones((3, 2))
a = np.arange(3)

We'll start by determining the shapes of the arrays:

- M.shape is (3,2)
- a.shape is (3,) 

Rule 1 says we must pad the shape of  `a`  with ones:

- a.shape is (3,2)
- b.shape becomes (1,3) 

By rule 2, the first dimension of `a` is then stretched to match that of `M`:

- M.shape is (3,2)
- a.shape becomes (3,3) 

Now we hit rule 3 — the final shapes do not match, so these two arrays are incompatible, as we can observe by attempting this operation:

In [113]:
M + a

ValueError: operands could not be broadcast together with shapes (3,2) (3,) 

#### Comparing arrays

You can compare arrays with individual values and with other arrays. Comparisons are performed element-wise. Such comparisons produce arrays of Boolean values in which each element’s `True` or `False` value indicates the comparison result:

In [41]:
numbers >= 13 # numbers = array([11, 12, 13, 14, 15])

array([False, False,  True,  True,  True])

The above implicitly used broadcasting!

In [42]:
numbers2 < numbers # numbers2 = array([ 1.1,  2.2,  3.3,  4.4,  5.5])

array([ True,  True,  True,  True,  True])

In [43]:
numbers == numbers2

array([False, False, False, False, False])

In [44]:
numbers == numbers

array([ True,  True,  True,  True,  True])

## Universal Functions

`NumPy` offers dozens of standalone ***universal functions*** (or `ufuncs`) that perform various element-wise operations (which means it applies the same operation to each element in the array). Each performs its task using one or two arrays or array-like (such as lists) arguments. Some of these functions are called when you use operators like `+` and `*` on arrays. Each returns a new array containing the results.

### The Slowness of Loops

Computation on `NumPy` arrays can be very fast or it can be very slow. The key to making it fast is to use ***vectorized operations***, generally implemented through `NumPy`'s universal functions (`ufuncs`). The relative sluggishness of `Python` generally manifests itself in situations where many small operations are being repeated; for instance, looping over arrays to operate on each element. For example, imagine we have an array of values and we’d like to compute the reciprocal of each. A straightforward approach might look like this:

In [117]:
def compute_reciprocals(values):
    output = np.empty(len(values))
    for i in range(len(values)):
        output[i] = 1.0 / values[i]
    return output

values = np.random.randint(1, 10, 5)
compute_reciprocals(values)

array([0.14285714, 0.14285714, 0.25      , 0.16666667, 0.2       ])

But if we measure the execution time of this code for a large input, we see that this operation is very slow:

In [118]:
big_array = np.random.randint(1, 10, 1_000_000)
%timeit compute_reciprocals(big_array)

1.22 s ± 9.99 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


It turns out that the bottleneck here is not the operations themselves but the type checking and function dispatches that `Python` must do at each cycle of the loop. Each time the reciprocal is computed, `Python` first examines the object’s type and does a dynamic lookup of the correct function to use for that type. If we were working in compiled code instead, this type of specification would be known before the code executed and the result could be computed much more efficiently.

**For many types of operations, `NumPy` provides a convenient interface into just this kind of statically typed, compiled routine. This is known as a vectorized operation**. For simple operations like the element-wise division here, vectorization is as simple as using `Python` arithmetic operators directly on the `array` object. This vectorized approach is designed to push the loop into the compiled layer that underlies `NumPy`, leading to much faster execution.

> Vectorization in `NumPy` refers to the practice of performing operations on entire arrays of data, rather than on individual elements. This is accomplished by applying an operation to the entire array, instead of looping through each element of the array one at a time.

In [119]:
print(1.0 / values)

[0.14285714 0.14285714 0.25       0.16666667 0.2       ]


Looking at the execution time for our big array, we see that it completes orders of magnitude faster than the `Python` loop:

In [120]:
%timeit (1.0 / big_array)

2.21 ms ± 58.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


Vectorized operations in `NumPy` are implemented via `ufuncs`, whose main purpose is to quickly execute repeated operations on values in `NumPy` array.

#### Exploring `NumPy`’s `Ufuncs`

Let's add two arrays with the same shape, using the `add()` universal function:

In [50]:
numbers2 = np.arange(1, 7) * 10 # array([10, 20, 30, 40, 50, 60])
np.add(numbers, numbers2) # equivalent to numbers + numbers2

array([11, 24, 39, 56, 75, 96])

#### Broadcasting with Universal Functions

Let's use the `multiply()` universal function to multiply every element of `numbers2` by the scalar value 5:

In [51]:
np.multiply(numbers2, 5) # equivalent to numbers2 * 5

array([ 50, 100, 150, 200, 250, 300])

Let's reshape `numbers2` into a 2-by-3 array, then multiply its values by a one-dimensional array of three elements:

In [52]:
numbers3 = numbers2.reshape(2, 3)
numbers4 = np.array([2, 4, 6])
numbers3, numbers4

(array([[10, 20, 30],
        [40, 50, 60]]),
 array([2, 4, 6]))

In [53]:
np.multiply(numbers3, numbers4) # Equivalent to numbers3 * numbers4

array([[ 20,  80, 180],
       [ 80, 200, 360]])

This works because `numbers4` has the same length as each row of `numbers3`, so `NumPy` can apply the multiply operation by treating `numbers4` as if it were the following array:

```python
array([[2, 4, 6],
       [2, 4, 6]])
```

If a universal function receives two arrays of different shapes that do not support broadcasting, a `ValueError` occurs. You can review the broadcasting rules at [https://numpy.org/doc/stable/user/basics.broadcasting.html](https://numpy.org/doc/stable/user/basics.broadcasting.html).

> Vectorization and `ufunc` functions are closely related to broadcasting in `NumPy`, as they are often used together to perform element-wise operations on arrays of different shapes. By combining vectorization, `ufunc` functions, and broadcasting, you can efficiently perform complex arithmetic operations on arrays in `NumPy`.

There are other special `ufunc`. Let's create an array and calculate the square root of its values, using the `sqrt` universal function:

In [121]:
numbers = np.array([1, 4, 9, 16, 25, 36])
np.sqrt(numbers)

array([1., 2., 3., 4., 5., 6.])

#### Create Your Own Vectorizing functions

To get good performance, we should try to avoid looping over elements in our vectors and matrices and instead use vectorized algorithms. The first step in converting a scalar algorithm to a vectorized algorithm is to ensure that the functions we write work with vector inputs.

In [122]:
def Theta(x):
    """
    Scalar implemenation of the Heaviside step function.
    """
    if x >= 0:
        return 1
    else:
        return 0

In [123]:
Theta_vec = np.vectorize(Theta)
Theta_vec(np.array([-3,-2,-1,0,1,2,3]))

array([0, 0, 0, 1, 1, 1, 1])

You can view the complete list, their descriptions and more information about universal functions at [https://numpy.org/doc/stable/reference/ufuncs.html](https://numpy.org/doc/stable/reference/ufuncs.html)

> See [here](https://www.labri.fr/perso/nrougier/from-python-to-numpy/#problem-vectorization) for more vectorization examples in different thinking levels.

### Type casting

Since `NumPy` arrays are **statically typed**, the type of an `array` does not change once created. But we can explicitly cast an array of some type to another using the `astype()` functions. This always creates a new array of a new type:

In [124]:
M = np.random.rand(5,5)
M.dtype

dtype('float64')

In [125]:
M2 = M.astype(np.int64)
M2

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]], dtype=int64)

See [https://scipy-lectures.org/intro/numpy/elaborate_arrays.html](https://scipy-lectures.org/intro/numpy/elaborate_arrays.html) for more details.

## File I/O

`NumPy` has its own binary format, not portable but with efficient I/O. This is useful when storing and reading back `array` data. Use the functions `numpy.save()` and `numpy.load()`.

In [126]:
np.save("random-matrix.npy", M)
M2 = np.load("random-matrix.npy")
M2

array([[0.63296042, 0.27197126, 0.2261708 , 0.1958696 , 0.9284412 ],
       [0.27045565, 0.16168923, 0.06145811, 0.91101895, 0.03550733],
       [0.52027297, 0.5984186 , 0.08252073, 0.80356072, 0.76129669],
       [0.20675944, 0.47959083, 0.28455249, 0.03627423, 0.4218361 ],
       [0.11976237, 0.47007539, 0.26740034, 0.51019782, 0.77871332]])

In summary:

To make the code faster using `NumPy`  

- In place operations: `a *= 3` instead of `a = 3*a`
- Use views instead of copies whenever possible
- Broadcasting: Use broadcasting to do operations on arrays
- Vectorizing `for` loops: Find tricks to avoid `for` loops using `NumPy` arrays.


**Python** objects:

- High-level objects: integers, floating-point
- Containers: lists ([costless](https://wiki.python.org/moin/TimeComplexity) append), dictionaries (fast lookup)
- Python lists are very general. They can contain any kind of object and are dynamically typed 
- However, they do not support mathematical functions such as matrix and dot multiplications. Implementing such functions for Python lists would not be very efficient because of the dynamic typing

**NumPy** provides:

- Extension package to `Python` for multi-dimensional arrays
- Numpy arrays are **statically typed** and **homogeneous**. The type of the elements is determined when the array is created
- Because of the static typing, fast implementation of mathematical functions such as multiplication and addition of `NumPy` arrays can be implemented in a compiled language (C and Fortran is used). Moreover, `Numpy` arrays are memory efficient